File: client/py/examples/read_latest.py
#  Copyright 2026 Synnax Labs, Inc.
#
#  Use of this software is governed by the Business Source License included in the file
#  licenses/BSL.txt.
#
#  As of the Change Date specified in that file, in accordance with the Business Source
#  License, use of this software will be governed by the Apache License, Version 2.0,
#  included in the file licenses/APL.txt.

"""
This example demonstrates how to read the latest N samples from Synnax channels. We'll:
1. Create an index channel for timestamps and a data channel for values
2. Write some sample data to these channels
3. Read the latest N samples using an iterator
4. Print the results

This pattern is useful for real-time monitoring applications where you need to access
the most recent data points.
"""

import numpy as np

import synnax as sy

# We've logged in via the command-line interface, so there's no need to provide
# credentials here. See https://docs.synnaxlabs.com/reference/client/quick-start.
client = sy.Synnax()

# Create an index channel that will be used to store our timestamps.
time_channel = client.channels.create(
    name="read_latest_time",
    is_index=True,
    data_type=sy.DataType.TIMESTAMP,
)

# Create a data channel that will be used to store our data.
data_channel = client.channels.create(
    name="read_latest_data",
    # We need to specify the index channel that will be used to store the timestamps for
    # this data channel.
    index=time_channel.key,
    data_type=sy.DataType.FLOAT32,
)

SAMPLE_COUNT = int(100)

# We'll start our write at the current time. This timestamp should be the same as or
# just before the first timestamp we write.
start = sy.TimeStamp.now()

# We'll end our write 100 seconds later
end = start + 2 * sy.TimeSpan.SECOND

# Generate linearly spaced int64 timestamps
stamps = np.linspace(start, end, SAMPLE_COUNT)

# Generate a sine wave with some noise as our data
data = np.sin(np.linspace(1, 10, SAMPLE_COUNT)) * 20 + np.random.randn(SAMPLE_COUNT)

# Write the data to the Synnax cluster through the channels. Note that we need to write
# to the index channel first, otherwise our write will fail.
time_channel.write(start, stamps)
data_channel.write(start, data)

print(client.read_latest(["daq_time", "ox_pt_1"], 1))

File: client/ts/examples/node/readLatest.js
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

/**
 * This example demonstrates how to read the latest N samples from Synnax channels. We'll:
 * 1. Create an index channel for timestamps and a data channel for values
 * 2. Write some sample data to these channels
 * 3. Read the latest N samples using an iterator
 * 4. Print the results
 *
 * This pattern is useful for real-time monitoring applications where you need to access
 * the most recent data points.
 */

import { DataType, Synnax, TimeSpan, TimeStamp } from "@synnaxlabs/client";

// Connect to a locally running, insecure Synnax cluster. If your connection parameters
// are different, enter them here. See https://docs.synnaxlabs.com/reference/client/quick-start.
const client = new Synnax({
  host: "localhost",
  port: 9090,
  username: "synnax",
  password: "seldon",
  secure: false,
});

// Create an index channel that will be used to store our timestamps.
const timeChannel = await client.channels.create({
  name: "read_latest_time",
  isIndex: true,
  dataType: DataType.TIMESTAMP,
});

// Create a data channel that will be used to store our data.
const dataChannel = await client.channels.create({
  name: "read_latest_data",
  // We need to specify the index channel that will be used to store the timestamps for
  // this data channel.
  index: timeChannel.key,
  dataType: DataType.FLOAT32,
});

const SAMPLE_COUNT = 100;

// We'll start our write at the current time. This timestamp should be the same as or
// just before the first timestamp we write.
const start = TimeStamp.now();

// We'll end our write 100 seconds later
const end = start.add(TimeSpan.seconds(2));

// Generate linearly spaced int64 timestamps
const stamps = BigInt64Array.from({ length: SAMPLE_COUNT }, (_, i) =>
  start.add(TimeSpan.milliseconds((i * 2000) / SAMPLE_COUNT)).valueOf(),
);

// Generate a sine wave with some noise as our data
const data = Float32Array.from(
  { length: SAMPLE_COUNT },
  (_, i) => Math.sin(1 + (i * 9) / SAMPLE_COUNT) * 20 + (Math.random() - 0.5),
);

// Write the data to the Synnax cluster through the channels. Note that we need to write
// to the index channel first, otherwise our write will fail.
await timeChannel.write(start, stamps);
await dataChannel.write(start, data);

console.log(await client.readLatest(["read_latest_time", "read_latest_data"], 1));

client.close();
